docs/reference/scripting.asciidoc

```diff
@@ -4,61 +4,72 @@
 [partintro]
 --
 With scripting, you can evaluate custom expressions in {es}. For example, you
-could use a script to return "script fields" as part of a search request or
-evaluate a custom score for a query.
+can use a script to return a computed value as a field or evaluate a custom
+score for a query.
 
-The default scripting language is <<modules-scripting-painless, `Painless`>>.
-Additional `lang` plugins enable you to run scripts written in other languages.
-Everywhere a script can be used, you can include a `lang` parameter
-to specify the language of the script.
+The default scripting language is <<modules-scripting-painless,Painless>>.
+Additional `lang` plugins are available to run scripts written in other
+languages. You can specify the language of the script anywhere that scripts run.
 
 [discrete]
-== General-purpose languages
+[[scripts-and-search-speed]]
+== Scripts and search speed
 
-These languages can be used for any purpose in the scripting APIs,
-and give the most flexibility.
+Scripts can't use {es}'s index structures or related optimizations, which can
+sometimes result in slower search speeds.
 
-[cols="<,<,<",options="header",]
-|=======================================================================
-|Language
-    |Sandboxed
-    |Required plugin
+If you often use scripts to transform indexed data, you can speed up search by
+making these changes during ingest instead. However, that often means slower
+index speeds.
 
-|<<modules-scripting-painless, `painless`>>
-    |yes
-    |built-in
+[discrete]
+[[modules-scripting-using-caching]]
+== Script caching
+
+All scripts are cached by default so that they only need to be recompiled
```

Comment:
There are a number of optimizations {es} performs to make using scripts as fast as possible.  One important optimization is a script cache.  The compiled form a script is placed in a cache so requests that use the same script do not incur a penalty compiling the script.

Cache sizing is important, your script cache should be large enough to hold all the scripts your users need to use concurrently.  

If you see a large number of script cache evictions and a rising number of compilations in [node stats](https://www.elastic.co/guide/en/elasticsearch/reference/current/cluster-nodes-stats.html), your cache may be too small.

Reply:
I'm incorporating this information into a new section for scripts, caching, and search speed.